base/pm/pm.go

```diff
@@ -498,6 +502,27 @@ func callback(cmd *Command, result *JobResult) {
 	}
 }
 
+func terminateAll(sig syscall.Signal, except ...string) {
+	jobsM.RLock()
+	defer jobsM.RUnlock()
+	for _, v := range jobs {
+		if utils.InString(except, v.Command().ID) {
+			continue
+		}
+
+		v.Terminate(sig)
+	}
+}
+
+//Shutdown kills all running processes.
+func Shutdown(except ...string) {
+	queue.Close()
+
+	terminateAll(syscall.SIGTERM, except...)
+	<-time.After(5 * time.Second)
```

Comment:
so we only allow 5 seconds before actually killing the processes ?
In the case of the node robot this won't be enough. The flow of the robot when it receives SIGTERM is:
- Stop accepting new http request
- Wait for all currently running request to finish
- Stop all services
  - stop the task queue processing
  -  wait for the running task is any, to finish

This all process can take more then 5 seconds specially if there are a lof of busy services

Reply:
You are right, I think the better way to do it is that after signalling each job that we actually wait for all of them to exit (not just time out) and then kill the ones that are still hanging in like 90 seconds. This way if all services exited immediately, we don't hog the system from shutting down. In other words, we only wait if we need to wait, otherwise, proceed with the shutdown.

Reply:
exactly what I had in mind, perfect.